Record and Playback - refactoring

HarshaNalluru opened the pull request:
Currently, the base recorder which allows recording the tests and playing them back is being duplicated across packages(for both node and browser that I and Caio have worked on).

Example -  recorder.ts for storage-queue package

In an effort to avoid the duplication of the recorder, this PR attempts to unify the recorder modules and provide a common utils library in the repo.

So, this new package "@azure/test-utils-recorder" can be added as a devDependency in an sdk (for example, storage-queue) to leverage the functionalities provided by the "recorder" to record the tests and generate recordings and to playback those recordings.

HarshaNalluru commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,344 @@
+const skip = [
+  // Abort
+  "browsers/aborter/recording_should_abort_after_aborter_timeout.json"
+];
```
Eventual plan is to remove the skip list - https://github.com/azure/azure-sdk-for-js/issues/4084

HarshaNalluru replied:
https://github.com/Azure/azure-sdk-for-js/pull/4336

HarshaNalluru commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,344 @@
+if (isPlayingBack) {
+  // Providing dummy values to avoid the error
+  env.ACCOUNT_NAME = "fakestorageaccount";
+  env.ACCOUNT_KEY = "aaaaa";
```
More work on ENV variables 
https://github.com/Azure/azure-sdk-for-js/issues/4284

HarshaNalluru commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,344 @@
+      new RegExp(env.ACCOUNT_NAME, "g"),
+      "fakestorageaccount"
+    );
+    if (env.ACCOUNT_KEY) {
```
More work on ENV variables
#4284

sadasant commented on sdk/test-utils/recorder/GUIDELINES.md:
```diff
@@ -0,0 +1,322 @@
+
+- `recorder` package assumes that the tests in the sdk are leveraging [mocha](https://mochajs.org/) and [rollup](https://rollupjs.org/guide/en/) (and [karma](https://karma-runner.github.io/latest/index.html) test runner for browser tests) as suggested by the [template](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/template/template) package in the repo.
+
+- Follow the below template for adding a new test. `before` and `after` sections are optional, `beforeEach` and `afterEach` sections are compulsory.
```
Is this correct? I believe it can be either. If I decide to just use before and after, I think I'm able to skip beforeEach and afterEach, same as if I decide to go with the *Each functions, then before and after are optional. At least that's how I've been doing things in keyvault.

HarshaNalluru replied:
[Discussed offline with @sadasant]
Decided to follow "starting the recorder in beforeeach and stopping in aftereach" method since it would allow recording a specific test separately.
I'll rephrase the statement in the Guidelines explaining how the recorder starts and stops which would clarify the need for beforeEach and afterEach sections.

ramya-rao-a commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,389 @@
+    env.STORAGE_CONNECTION_STRING = `DefaultEndpointsProtocol=https;AccountName=${
+      env.ACCOUNT_NAME
+    };AccountKey=${env.ACCOUNT_KEY};EndpointSuffix=core.windows.net`;
+  }
```
This is related to storage and needs to be removed.

HarshaNalluru replied:
Added

ramya-rao-a commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,389 @@
+  protected readonly filepath: string;
+  public uniqueTestInfo: TestInfo = { uniqueName: {}, newDate: {} };
+
+  constructor(runtime: string, testHierarchy: string, testTitle: string, ext: string) {
```
`runtime` -> `platform`

Consider having the type has either an enum or `"node" | "browser"`

HarshaNalluru replied:
Changed `testHierarchy` ->  `testSuiteTitle`, `runtime` -> `platform`, 
Union type `platform: "node" | "browsers"` and removed `ext` argument.

ramya-rao-a commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,389 @@
+      "/recording_" +
+      this.formatPath(testTitle) +
+      "." +
+      ext;
```
consider using `path.join(runtime, this.formatPath(testHeirarchy), ....)`

HarshaNalluru replied:
`path` module is not available in the browsers.

jonathandturner replied:
Can we test for node?

HarshaNalluru replied:
This code path is accessed in both `node` and `browser` runtimes. 
`path` module is not available in the browsers, that's the reason we are not using the `path` module.

ramya-rao-a commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,389 @@
+      updatedRecording = map(updatedRecording);
+    }
+
+    // Handling storage environment variables separately
```
Same as https://github.com/Azure/azure-sdk-for-js/pull/4281/files?file-filters%5B%5D=.json&file-filters%5B%5D=.ts#r308844665, add a TODO comment

HarshaNalluru replied:
Added

ramya-rao-a commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,389 @@
+  }
+
+  public playback(filePath: string): void {
+    let path = require("path");
```
Add a comment as to why we need to dynamically import this

HarshaNalluru replied:
Added

ramya-rao-a commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,389 @@
+    });
+
+    file.write(
+      importNock + "\n" + "module.exports.testInfo = " + JSON.stringify(this.uniqueTestInfo) + "\n"
```
Move this entire string manipulation into the `importNock` constant and rename it to `importNockStatement`

HarshaNalluru replied:
Done

ramya-rao-a commented on sdk/test-utils/recorder/src/delay.ts:
```diff
@@ -0,0 +1,15 @@
+ * @param {number} milliseconds The number of milliseconds to be delayed.
+ * @returns {Promise<T>} Resolved promise
+ */
+export function delay(milliseconds: number): Promise<void> | null {
```
Why a separate file for this? Consider moving this too `utils.ts` file

HarshaNalluru replied:
Moved

ramya-rao-a commented on sdk/test-utils/recorder/src/utils.ts:
```diff
@@ -0,0 +1,78 @@
+  return env.TEST_MODE === "record";
+}
+
+export function isPlayingBack() {
```
isRecording -> isRecordMode
isPlayingBack -> isPlaybackMode

HarshaNalluru replied:
Changed

ramya-rao-a commented on sdk/test-utils/recorder/src/recorder.ts:
```diff
@@ -0,0 +1,115 @@
+ * @export
+ * An interface that allows recording and playback capabilities for the tests in Azure TS SDKs.
+ */
+export interface Recorder {
```
I find it slightly confusing that the interface `Recorder` interface is not really for the real recorders like the `NockRecorder` but is for a set of functions.

HarshaNalluru replied:
`NockRecorder` is not exposed to the users and is internal to the recorder package.

If you still feel like it might be confusing, `Recorder` can be renamed to `RecorderHandler` to differentiate it from `NockRecorder`/`BaseRecorder`.

What say?

jonathandturner replied:
If only one `Recorder` type is exposed, I think it's fine to keep the name.

HarshaNalluru commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,414 @@
+// Nise module does not have a native implementation of record/playback like Nock does
+// This class overrides requests' 'open', 'send' and 'onreadystatechange' functions, adding our own code to them to deal with requests
+export class NiseRecorder extends BaseRecorder {
+  private readonly sasQueryParameters = ["se", "sig", "sp", "spr", "srt", "ss", "st", "sv"];
```
This is to filter sasQueryParameters in the browser tests for storage sdks.
Code related to this doesn't affect if such query params are not present.

Will work in a future PR to update this make the recorder more generic.

/cc - @sadasant

jonathandturner replied:
These hardcoded parameters should be passed into the library rather than part of the library itself. Different services may have their own query parameters, potentially.

HarshaNalluru replied:
I agree.

HarshaNalluru replied:
Addressed below - https://github.com/Azure/azure-sdk-for-js/pull/4281#discussion_r316844252

HarshaNalluru replied:
Addressed below - https://github.com/Azure/azure-sdk-for-js/pull/4281#discussion_r316844252

jonathandturner commented on sdk/test-utils/recorder/GUIDELINES.md:
````diff
@@ -0,0 +1,333 @@
+  For example
+
+  ```typescript
+  describe("Aborter", () => {
````
I thought we weren't using this name

HarshaNalluru replied:
Hmm. We haven't updated the title in storage test-suites.

I'll update this to "Some Random Test Suite".

jonathandturner commented on sdk/test-utils/recorder/GUIDELINES.md:
```diff
@@ -0,0 +1,333 @@
+
+- The above methods can be called from the `before` section so that the tests can leverage the environment variables.
+
+- Currently, the environment variables for storage-packages are managed by the recorder. `ACCOUNT_NAME` is replaced with `"fakestorageaccount"`, secret part of `ACCOUNT_SAS` and `ACCOUNT_KEY` are replaced with `"aaaaa"`.
```
Storage-specific stuff should live outside of the recorder package.

HarshaNalluru replied:
Removed

jonathandturner commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,414 @@
+    env.ACCOUNT_KEY = "aaaaa";
+    env.ACCOUNT_SAS = "aaaaa";
+    env.STORAGE_CONNECTION_STRING = `DefaultEndpointsProtocol=https;AccountName=${
+      env.ACCOUNT_NAME
```
Storage-specific functionality should be outside of the recorder.  I see the TODO, but it'd be good to make sure we're landing something general first, and then we can land PRs for storage and kv to use it.

HarshaNalluru replied:
Done

jonathandturner commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,414 @@
+      responseHeaders[key] = value;
+    }
+
+    // We're not storing SAS Query Parameters because they may contain sensitive information
```
It seems like the service should send you the additional set of parameters that may contain sensitive information

HarshaNalluru replied:
Updated

jonathandturner commented on sdk/test-utils/recorder/src/baseRecorder.ts:
```diff
@@ -0,0 +1,414 @@
+          const reqStateChange = req.onreadystatechange;
+          req.onreadystatechange = function() {
+            // Record the request once the response is obtained
+            if (req.readyState === 4) {
```
Can you say more about what readyState is, and what in particular 4 is? Could this be an enum instead?

HarshaNalluru replied:
Added some comments stating what `readyState` refers to.

jonathandturner replied:
I mean more the hard-coded 4. Why is it 4 and not 3 or 5?

HarshaNalluru replied:
More info - https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState
![image](https://user-images.githubusercontent.com/10452642/63553698-1da41b00-c4f0-11e9-8ea8-1ed622d4f442.png)




Added this comment - `readyState = 4 refers to the completion of the operation.`. Should I talk more about it?

jonathandturner replied:
Ah okay, so there isn't an official enumeration, just the number. It's fine to leave it as is, but you may want to link to the documentation for people who try to read the code later

HarshaNalluru commented:
@jonathandturner  and @sadasant 

**Tested the new `@azure/test-utils-recorder` with keyvault-secrets package**
 - Imported the new Keyvault into a branch and updated keyvault-secrets package to leverage the `@azure/test-utils-recorder`
- Branch (as a PR to view the changes easily) - https://github.com/HarshaNalluru/azure-sdk-for-js/pull/15
	
**Findings/comments/questions/suggestions**
- unit-test:browser is "echo skipped" - Is there a reason behind skipping the unit tests for the  browser ?
- Deleted the recordings for node tests
- Generated the recordings for node tests - Successful
- Playback is also successful for node tests - Successful
- Some of the tests have same title and are in the same test-suite - this needs to be updated
- I've observed that we need to set env variables in the .env file and using `set var=value`
  We can update the code to import env variables from .env file using dotenv all the time and reduce the complexity of doing  `set var=value` each time we run the tests.
- One suggestion to update the test:node command to 
`"test:node": "npm run clean && npm run build:test && npm run integration-test:node",` and use it instead of having to run both build:test and integration:test
[ I kept forgetting that I need to run `build:test` before running `integration-test:node"`]

KarishmaGhiya commented on sdk/test-utils/recorder/package.json:
```diff
@@ -0,0 +1,82 @@
+    "integration-test:node": "echo skipped",
+    "integration-test": "npm run integration-test:node && npm run integration-test:browser",
+    "lint:fix": "eslint \"src/**/*.ts\" \"test/**/*.ts\" -c ../../.eslintrc.json --fix --fix-type [problem,suggestion]",
+    "lint": "eslint -c ../../.eslintrc.json src test --ext .ts -f node_modules/eslint-detailed-reporter/lib/detailed.js -o template-lintReport.html || exit 0",
```
Can you submit another PR and change the lint step to : 
 "lint": "eslint -c ../../.eslintrc.json src test --ext .ts -f html -o recorder-lintReport.html || exit 0",

HarshaNalluru replied:
Thanks

HarshaNalluru replied:
https://github.com/Azure/azure-sdk-for-js/pull/4886
